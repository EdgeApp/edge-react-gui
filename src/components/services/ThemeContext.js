// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
