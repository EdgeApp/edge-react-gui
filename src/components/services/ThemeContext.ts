import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
