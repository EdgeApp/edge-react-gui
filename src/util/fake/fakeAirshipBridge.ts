import { AirshipBridge } from 'react-native-airship'

export const fakeAirshipBridge: AirshipBridge<any> = {
  resolve: value => undefined,
  // eslint-disable-next-line node/handle-callback-err
  reject: error => undefined,
  remove: () => undefined,
  on: (name, callback) => () => undefined,
  onResult: callback => undefined
}
