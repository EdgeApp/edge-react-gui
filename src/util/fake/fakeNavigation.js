// @flow

import { type NavigationProp } from '../../types/routerTypes.js'

export const fakeNavigation: NavigationProp<any> = {
  addListener(event, callback) {
    return () => undefined
  },
  isFocused() {
    return true
  },

  navigate(name, params) {},
  push(name, params) {},
  replace(name, params) {},
  setParams(params) {},
  jumpTo(name, params) {},

  goBack() {},
  pop() {},
  popToTop() {},

  closeDrawer() {},
  openDrawer() {},
  toggleDrawer() {},

  state: {},
  currentScene: {}
}
