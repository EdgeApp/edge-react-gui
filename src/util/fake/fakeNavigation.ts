import { NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener(event, callback) {
    return () => undefined
  },
  isFocused() {
    return true
  },

  navigate(name, params) {},
  push(name, params) {},
  replace(name, params) {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popToTop() {},

  closeDrawer() {},
  openDrawer() {},
  toggleDrawer() {},

  state: {}
}
