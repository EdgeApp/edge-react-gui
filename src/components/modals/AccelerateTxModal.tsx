import { EdgeCurrencyWallet, EdgeDenomination, EdgeTransaction } from 'edge-core-js'
import React, { PureComponent } from 'react'
import { Text, View } from 'react-native'
import { AirshipBridge } from 'react-native-airship'

import s from '../../locales/strings'
import { Slider } from '../../modules/UI/components/Slider/Slider'
import { getDisplayDenominationFromState, getExchangeDenominationFromState } from '../../selectors/DenominationSelectors'
import { connect } from '../../types/reactRedux'
import { GuiExchangeRates } from '../../types/types'
import { convertTransactionFeeToDisplayFee } from '../../util/utils'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../services/ThemeContext'
import { ModalFooter, ModalMessage, ModalTitle } from '../themed/ModalParts'
import { ThemedModal } from '../themed/ThemedModal'
import { Tile } from '../tiles/Tile'

interface OwnProps {
  acceleratedTx: EdgeTransaction
  bridge: AirshipBridge<EdgeTransaction | null>
  replacedTx: EdgeTransaction
  wallet: EdgeCurrencyWallet
}
interface StateProps {
  exchangeRates: GuiExchangeRates
}
interface DispatchProps {
  getDisplayDenomination: (pluginId: string, currencyCode: string) => EdgeDenomination
  getExchangeDenomination: (pluginId: string, currencyCode: string) => EdgeDenomination
}
type Props = OwnProps & StateProps & ThemeProps & DispatchProps

interface State {
  error?: Error
  status: 'confirming' | 'sending' | 'sent'
  mounted: boolean
}

export class AccelerateTxModalComponent extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {
      error: undefined,
      status: 'confirming',
      mounted: true
    }
  }

  signBroadcastAndSave = async () => {
    const { acceleratedTx, wallet } = this.props

    this.setState({ status: 'sending' })

    try {
      // Sign, broadcast, and save the accelerated transaction
      const signedTx = await wallet.signTx(acceleratedTx)
      await wallet.broadcastTx(signedTx)
      await wallet.saveTx(signedTx)

      if (this.state.mounted) {
        this.setState({ status: 'sent' })
      }
      this.closeModal(signedTx)
    } catch (error: any) {
      if (this.state.mounted) {
        this.closeModal(null, error)
      }
    }
  }

  closeModal = (signedTx: EdgeTransaction | null, err?: Error) => {
    this.setState({
      mounted: false
    })
    if (err != null) this.props.bridge.reject(err)
    else this.props.bridge.resolve(signedTx)
  }

  handleCancel = () => {
    this.closeModal(null)
  }

  handleConfirmation = async () => {
    await this.signBroadcastAndSave()
  }

  getTxFeeDisplay = (edgeTransaction: EdgeTransaction): string => {
    const { exchangeRates, wallet, getDisplayDenomination, getExchangeDenomination } = this.props

    const feeDisplayDenomination = getDisplayDenomination(wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const feeDefaultDenomination = getExchangeDenomination(wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const transactionFee = convertTransactionFeeToDisplayFee(wallet, exchangeRates, edgeTransaction, feeDisplayDenomination, feeDefaultDenomination)

    const feeSyntax = `${transactionFee.cryptoSymbol ?? ''} ${transactionFee.cryptoAmount} (${transactionFee.fiatSymbol ?? ''} ${transactionFee.fiatAmount})`

    return feeSyntax
  }

  render() {
    const { acceleratedTx, bridge, replacedTx, theme } = this.props
    const { error, status } = this.state

    const styles = getStyles(theme)

    const oldFee = this.getTxFeeDisplay(replacedTx)
    const newFee = this.getTxFeeDisplay(acceleratedTx)

    const isSending = status === 'sending'

    return (
      <ThemedModal bridge={bridge} onCancel={this.handleCancel}>
        <ModalTitle>{s.strings.transaction_details_accelerate_transaction_header}</ModalTitle>
        <ModalMessage>{s.strings.transaction_details_accelerate_transaction_instructional}</ModalMessage>
        <View style={styles.container}>
          <Tile type="static" title={s.strings.transaction_details_accelerate_transaction_old_fee_title} body={oldFee} />
          {newFee == null ? null : <Tile type="static" title={s.strings.transaction_details_accelerate_transaction_new_fee_title} body={newFee} />}
        </View>
        {error == null ? null : (
          <View style={styles.error}>
            <Text style={styles.errorText} numberOfLines={3}>
              {error.message}
            </Text>
          </View>
        )}
        <View style={styles.container}>
          <Slider
            disabled={isSending || !!error}
            onSlidingComplete={this.handleConfirmation}
            showSpinner={isSending}
            disabledText={s.strings.transaction_details_accelerate_transaction_slider_disabled}
          />
        </View>
        <ModalFooter onPress={this.handleCancel} />
      </ThemedModal>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  loadingContianer: {
    height: theme.rem(20)
  },
  error: {
    marginVertical: theme.rem(0.5)
  },
  errorText: {
    textAlign: 'center',
    color: theme.negativeText
  },
  loading: {
    flex: 1,
    alignSelf: 'center'
  },
  container: {
    maxHeight: theme.rem(20),
    marginVertical: theme.rem(0.5)
  }
}))

export const AccelerateTxModal = connect<StateProps, DispatchProps, OwnProps>(
  (state, ownProps) => ({
    exchangeRates: state.exchangeRates
  }),
  dispatch => ({
    getDisplayDenomination(pluginId: string, currencyCode: string) {
      return dispatch(getDisplayDenominationFromState(pluginId, currencyCode))
    },
    getExchangeDenomination(pluginId: string, currencyCode: string) {
      return dispatch(getExchangeDenominationFromState(pluginId, currencyCode))
    }
  })
)(withTheme(AccelerateTxModalComponent))
