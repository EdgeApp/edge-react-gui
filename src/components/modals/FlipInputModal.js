// @flow

import { bns } from 'biggystring'
import { type EdgeDenomination, asMaybeNoAmountSpecifiedError } from 'edge-core-js'
import * as React from 'react'
import { TouchableWithoutFeedback, View } from 'react-native'
import { type AirshipBridge } from 'react-native-airship'
import FontAwesomeIcon from 'react-native-vector-icons/FontAwesome'
import { sprintf } from 'sprintf-js'

import { updateMaxSpend, updateTransactionAmount } from '../../actions/SendConfirmationActions.js'
import { MINIMUM_DEVICE_HEIGHT } from '../../constants/constantSettings.js'
import { getSpecialCurrencyInfo } from '../../constants/WalletAndCurrencyConstants.js'
import { formatNumber } from '../../locales/intl.js'
import s from '../../locales/strings.js'
import { getDisplayDenomination, getExchangeDenomination } from '../../selectors/DenominationSelectors.js'
import { getExchangeRate } from '../../selectors/WalletSelectors.js'
import { deviceHeight } from '../../theme/variables/platform.js'
import { connect } from '../../types/reactRedux.js'
import type { GuiCurrencyInfo } from '../../types/types.js'
import { getCurrencyIcon } from '../../util/CurrencyInfoHelpers.js'
import { getWalletFiat, getWalletName } from '../../util/CurrencyWalletHelpers.js'
import { convertTransactionFeeToDisplayFee, DECIMAL_PRECISION, getDenomFromIsoCode, truncateDecimals } from '../../util/utils.js'
import { ExchangeRate } from '../common/ExchangeRate.js'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'
import { Card } from '../themed/Card'
import { EdgeText } from '../themed/EdgeText.js'
import { type ExchangedFlipInputAmounts, ExchangedFlipInput } from '../themed/ExchangedFlipInput'
import { FiatText } from '../themed/FiatText.js'
import { MiniButton } from '../themed/MiniButton.js'
import { ThemedModal } from '../themed/ThemedModal.js'

type OwnProps = {
  bridge: AirshipBridge<void>,
  walletId: string,
  currencyCode: string,
  onFeesChange: () => void
}

type StateProps = {
  // Balance
  balanceCrypto: string,

  // FlipInput
  flipInputHeaderText: string,
  flipInputHeaderLogo: string,
  primaryInfo: GuiCurrencyInfo,
  secondaryInfo: GuiCurrencyInfo,
  fiatPerCrypto: string,
  overridePrimaryExchangeAmount: string,
  forceUpdateGuiCounter: number,

  // Fees
  feeCurrencyCode: string,
  feeDisplayDenomination: EdgeDenomination,
  feeNativeAmount: string,
  feeAmount: string,
  feeStyle?: string,

  // Error
  errorMessage?: string
}

type DispatchProps = {
  updateMaxSpend: (walletId: string, currencyCode: string) => void,
  updateTransactionAmount: (nativeAmount: string, exchangeAmount: string, walletId: string, currencyCode: string) => void
}

type State = {
  overridePrimaryExchangeAmount: string,
  forceUpdateGuiCounter: number,
  errorMessage?: string
}

type Props = OwnProps & StateProps & DispatchProps & ThemeProps

class FlipInputModalComponent extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      overridePrimaryExchangeAmount: props.overridePrimaryExchangeAmount,
      forceUpdateGuiCounter: 0
    }
  }

  handleCloseModal = () => this.props.bridge.resolve()

  handleFeesChange = () => {
    this.handleCloseModal()
    this.props.onFeesChange()
  }

  handleExchangeAmountChange = ({ nativeAmount, exchangeAmount }: ExchangedFlipInputAmounts) => {
    const { walletId, currencyCode, updateTransactionAmount } = this.props
    updateTransactionAmount(nativeAmount, exchangeAmount, walletId, currencyCode)
  }

  handleAmountChangeError = (errorMessage?: string) => this.setState({ errorMessage })

  componentDidUpdate(prevProps: Props) {
    if (this.props.forceUpdateGuiCounter !== this.state.forceUpdateGuiCounter) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        overridePrimaryExchangeAmount: this.props.overridePrimaryExchangeAmount,
        forceUpdateGuiCounter: this.props.forceUpdateGuiCounter
      })
    }
  }

  handleSendMaxAmount = () => this.props.updateMaxSpend(this.props.walletId, this.props.currencyCode)

  renderErrorMessge = () => {
    const { errorMessage = this.state.errorMessage, theme } = this.props
    const styles = getStyles(theme)
    const opacity = errorMessage == null ? 0 : 1
    return (
      <EdgeText numberOfLines={1} style={[styles.exchangeRateErrorText, { opacity }]}>
        {errorMessage == null ? ' ' : errorMessage.split('\n')[0]}
      </EdgeText>
    )
  }

  renderExchangeRates = () => {
    const { primaryInfo, secondaryInfo, fiatPerCrypto, theme } = this.props
    const styles = getStyles(theme)

    return (
      <View style={styles.rateBalanceContainer}>
        <EdgeText style={styles.secondaryTitle}>{s.strings.string_rate}</EdgeText>
        <ExchangeRate primaryInfo={primaryInfo} secondaryInfo={secondaryInfo} secondaryDisplayAmount={fiatPerCrypto} style={styles.rateBalanceText} />
      </View>
    )
  }

  renderBalance = () => {
    const { balanceCrypto, primaryInfo, secondaryInfo, theme } = this.props
    const styles = getStyles(theme)
    const { multiplier, name } = primaryInfo.displayDenomination
    const balance = `${formatNumber(bns.div(balanceCrypto, multiplier, DECIMAL_PRECISION))} ${name} `
    return (
      <View style={styles.rateBalanceContainer}>
        <EdgeText style={styles.secondaryTitle}>{s.strings.send_confirmation_balance}</EdgeText>
        <EdgeText style={styles.rateBalanceText}>
          {balance}
          <FiatText
            nativeCryptoAmount={balanceCrypto}
            cryptoCurrencyCode={primaryInfo.exchangeCurrencyCode}
            isoFiatCurrencyCode={secondaryInfo.exchangeCurrencyCode}
            cryptoExchangeMultiplier={primaryInfo.exchangeDenomination.multiplier}
            parenthesisEnclosed
          />
        </EdgeText>
      </View>
    )
  }

  renderFlipInput = () => {
    const { flipInputHeaderText, flipInputHeaderLogo, primaryInfo, secondaryInfo, fiatPerCrypto } = this.props
    const { overridePrimaryExchangeAmount } = this.state
    return (
      <Card marginRem={0}>
        <ExchangedFlipInput
          headerText={flipInputHeaderText}
          headerLogo={flipInputHeaderLogo}
          primaryCurrencyInfo={{ ...primaryInfo }}
          secondaryCurrencyInfo={{ ...secondaryInfo }}
          exchangeSecondaryToPrimaryRatio={fiatPerCrypto}
          overridePrimaryExchangeAmount={overridePrimaryExchangeAmount}
          forceUpdateGuiCounter={0}
          onExchangeAmountChanged={this.handleExchangeAmountChange}
          onError={this.handleAmountChangeError}
          onNext={this.handleCloseModal}
          keyboardVisible={false}
          isFocus
          isFiatOnTop={bns.eq(overridePrimaryExchangeAmount, '0')}
        />
        {getSpecialCurrencyInfo(this.props.currencyCode).noMaxSpend !== true ? (
          <MiniButton alignSelf="center" label={s.strings.string_max_cap} marginRem={[1.2, 0, 0]} onPress={this.handleSendMaxAmount} />
        ) : null}
      </Card>
    )
  }

  renderFees = () => {
    const { feeAmount, feeCurrencyCode, feeDisplayDenomination, feeNativeAmount, feeStyle, primaryInfo, secondaryInfo, theme } = this.props
    const truncatedFeeAmount = truncateDecimals(feeAmount, 6, false)
    const feeCryptoText = `${truncatedFeeAmount} ${feeDisplayDenomination.name} `
    const styles = getStyles(theme)
    const feeTextStyle = feeStyle === 'dangerText' ? styles.feeTextDanger : feeStyle === 'warningText' ? styles.feeTextWarning : styles.feeTextDefault
    return (
      <View style={styles.feeContainer}>
        <View style={styles.feeTitleContainer}>
          <EdgeText style={styles.primaryTitle}>{s.strings.string_fee}</EdgeText>
          <FontAwesomeIcon name="edit" style={styles.feeIcon} size={theme.rem(0.75)} />
        </View>
        <EdgeText style={feeTextStyle}>
          {feeCryptoText}
          <FiatText
            nativeCryptoAmount={feeNativeAmount}
            cryptoCurrencyCode={feeCurrencyCode}
            isoFiatCurrencyCode={secondaryInfo.exchangeCurrencyCode}
            cryptoExchangeMultiplier={primaryInfo.exchangeDenomination.multiplier}
            parenthesisEnclosed
          />
        </EdgeText>
      </View>
    )
  }

  render() {
    const { theme } = this.props
    const styles = getStyles(theme)
    return (
      <ThemedModal bridge={this.props.bridge} onCancel={this.handleCloseModal}>
        {/* Extra view needed here to fullscreen the modal on small devices */}
        <View style={styles.hackContainer}>
          <View style={styles.flipInput}>{this.renderFlipInput()}</View>
          <TouchableWithoutFeedback onPress={this.handleFeesChange} style={styles.content}>
            <View>
              {this.renderFees()}
              {this.renderExchangeRates()}
              {this.renderBalance()}
              {this.renderErrorMessge()}
            </View>
          </TouchableWithoutFeedback>
        </View>
      </ThemedModal>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  hackContainer: {
    flex: deviceHeight <= MINIMUM_DEVICE_HEIGHT ? 1 : 0
  },
  flipInput: {
    justifyContent: 'flex-start'
  },
  content: {
    justifyContent: 'flex-end'
  },
  headerContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  headerMaxAmountText: {
    color: theme.textLink
  },
  primaryTitle: {
    color: theme.secondaryText
  },
  secondaryTitle: {
    flex: 1,
    fontSize: theme.rem(0.75),
    color: theme.secondaryText
  },
  rateBalanceContainer: {
    flexDirection: 'row'
  },
  exchangeRateErrorText: {
    fontSize: theme.rem(0.75),
    color: theme.dangerText
  },
  rateBalanceText: {
    fontSize: theme.rem(0.75)
  },
  feeContainer: {
    flexDirection: 'row',
    marginTop: theme.rem(0.5),
    marginBottom: theme.rem(1)
  },
  feeTitleContainer: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center'
  },
  feeTextDefault: {
    color: theme.primaryText
  },
  feeTextWarning: {
    color: theme.warningText
  },
  feeTextDanger: {
    color: theme.dangerText
  },
  feeIcon: {
    color: theme.iconTappable,
    marginLeft: theme.rem(0.5)
  }
}))

export const FlipInputModal = connect<StateProps, DispatchProps, OwnProps>(
  (state, ownProps) => {
    const { walletId, currencyCode } = ownProps
    const wallet = state.core.account.currencyWallets[walletId]
    const name = getWalletName(wallet)
    const { fiatCurrencyCode, isoFiatCurrencyCode } = getWalletFiat(wallet)
    const { symbolImageDarkMono } = getCurrencyIcon(wallet.currencyInfo.currencyCode, currencyCode)

    // Denominations
    const cryptoDenomination = getDisplayDenomination(state, wallet.currencyInfo.pluginId, currencyCode)
    const cryptoExchangeDenomination = getExchangeDenomination(state, wallet.currencyInfo.pluginId, currencyCode)
    const fiatDenomination = getDenomFromIsoCode(fiatCurrencyCode)

    // FlipInput
    const fiatPerCrypto = getExchangeRate(state, currencyCode, isoFiatCurrencyCode)

    const primaryInfo = {
      displayCurrencyCode: currencyCode,
      displayDenomination: cryptoDenomination,
      exchangeCurrencyCode: cryptoExchangeDenomination.name,
      exchangeDenomination: cryptoExchangeDenomination
    }

    const secondaryInfo = {
      displayCurrencyCode: fiatCurrencyCode,
      displayDenomination: fiatDenomination,
      exchangeCurrencyCode: isoFiatCurrencyCode,
      exchangeDenomination: fiatDenomination
    }

    const { forceUpdateGuiCounter, nativeAmount } = state.ui.scenes.sendConfirmation
    const overridePrimaryExchangeAmount = bns.div(nativeAmount, primaryInfo.exchangeDenomination.multiplier, DECIMAL_PRECISION)

    // Fees
    const feeDisplayDenomination = getDisplayDenomination(state, wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const feeDefaultDenomination = getExchangeDenomination(state, wallet.currencyInfo.pluginId, wallet.currencyInfo.currencyCode)
    const transactionFee = convertTransactionFeeToDisplayFee(
      wallet,
      state.exchangeRates,
      state.ui.scenes.sendConfirmation.transaction,
      feeDisplayDenomination,
      feeDefaultDenomination
    )

    // Error
    const error = state.ui.scenes.sendConfirmation.error
    let errorMessage
    if (error && error.message !== 'broadcastError' && error.message !== 'transactionCancelled' && asMaybeNoAmountSpecifiedError(error) == null) {
      errorMessage = error.message
    }

    return {
      // Balances
      balanceCrypto: wallet.balances[currencyCode],

      // FlipInput
      flipInputHeaderText: sprintf(s.strings.send_from_wallet, name),
      flipInputHeaderLogo: symbolImageDarkMono,
      primaryInfo,
      secondaryInfo,
      fiatPerCrypto: fiatPerCrypto ?? '0',
      overridePrimaryExchangeAmount,
      forceUpdateGuiCounter,

      // Fees
      feeCurrencyCode: wallet.currencyInfo.currencyCode,
      feeDisplayDenomination,
      feeNativeAmount: transactionFee.nativeCryptoAmount,
      feeAmount: transactionFee.cryptoAmount,
      feeStyle: transactionFee.fiatStyle,

      // Error
      errorMessage
    }
  },
  dispatch => ({
    updateMaxSpend(walletId: string, currencyCode: string) {
      dispatch(updateMaxSpend(walletId, currencyCode))
    },
    updateTransactionAmount(nativeAmount: string, exchangeAmount: string, walletId: string, currencyCode: string) {
      dispatch(updateTransactionAmount(nativeAmount, exchangeAmount, walletId, currencyCode))
    }
  })
)(withTheme(FlipInputModalComponent))
