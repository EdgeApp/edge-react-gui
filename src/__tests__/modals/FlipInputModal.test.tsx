import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

import { FlipInputModalComponent } from '../../components/modals/FlipInputModal'
import { getTheme } from '../../components/services/ThemeContext'
import { fakeAirshipBridge } from '../../util/fake/fakeAirshipBridge'
import { fakeNavigation } from '../../util/fake/fakeNavigation'

describe('FlipInputModalComponent', () => {
  it('should render with loading props', () => {
    const renderer = createRenderer()

    const fakeWallet: any = {}

    const actual = renderer.render(
      <FlipInputModalComponent
        bridge={fakeAirshipBridge}
        navigation={fakeNavigation}
        walletId="myWallet"
        wallet={fakeWallet}
        currencyCode="BTC"
        onFeesChange={() => undefined}
        balanceCrypto="10000"
        flipInputHeaderText="Exchange Header"
        primaryInfo={
          {
            tokenId: undefined,
            displayCurrencyCode: 'BTC',
            exchangeCurrencyCode: 'BTC',
            displayDenomination: { multiplier: '100000000000', name: 'BTC' },
            exchangeDenomination: { multiplier: '100000000000', name: 'BTC' }
          } as any
        }
        secondaryInfo={
          {
            displayCurrencyCode: 'BTC',
            exchangeCurrencyCode: 'BTC',
            displayDenomination: {
              name: 'Bitcoin',
              multiplier: '1'
            },
            exchangeDenomination: {
              name: 'Bitcoin',
              multiplier: '1'
            }
          } as any
        }
        fiatPerCrypto="1000"
        overridePrimaryExchangeAmount="0"
        forceUpdateGuiCounter={123}
        pluginId="Wyre"
        feeDisplayDenomination={{ multiplier: '100000000000', name: 'BTC' }}
        feeNativeAmount="1"
        feeAmount="1"
        updateMaxSpend={(walletId, currencyCode) => undefined}
        updateTransactionAmount={(nativeAmount, exchangeAmount, walletId, currencyCode) => undefined}
        theme={getTheme()}
      />
    )

    expect(actual).toMatchSnapshot()
  })
})
