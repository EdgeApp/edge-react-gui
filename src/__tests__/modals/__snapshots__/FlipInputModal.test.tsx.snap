// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlipInputModalComponent should render with loading props 1`] = `
<ThemedModal
  bridge={
    {
      "on": [Function],
      "onResult": [Function],
      "reject": [Function],
      "remove": [Function],
      "resolve": [Function],
    }
  }
  onCancel={[Function]}
>
  <View
    style={
      {
        "flex": 0,
      }
    }
  >
    <View
      style={
        {
          "justifyContent": "flex-start",
        }
      }
    >
      <WithTheme(CardComponent)>
        <ExchangedFlipInput
          exchangeSecondaryToPrimaryRatio="1000"
          headerText="Exchange Header"
          isEditable={true}
          isFiatOnTop={true}
          isFocus={true}
          keyboardVisible={false}
          onError={[Function]}
          onExchangeAmountChanged={[Function]}
          onNext={[Function]}
          overridePrimaryExchangeAmount="0"
          primaryCurrencyInfo={
            {
              "displayCurrencyCode": "BTC",
              "displayDenomination": {
                "multiplier": "100000000000",
                "name": "BTC",
              },
              "exchangeCurrencyCode": "BTC",
              "exchangeDenomination": {
                "multiplier": "100000000000",
                "name": "BTC",
              },
              "tokenId": undefined,
            }
          }
          secondaryCurrencyInfo={
            {
              "displayCurrencyCode": "BTC",
              "displayDenomination": {
                "multiplier": "1",
                "name": "Bitcoin",
              },
              "exchangeCurrencyCode": "BTC",
              "exchangeDenomination": {
                "multiplier": "1",
                "name": "Bitcoin",
              },
            }
          }
        />
        <MiniButton
          alignSelf="center"
          label="MAX"
          marginRem={
            [
              1.2,
              0,
              0,
            ]
          }
          onPress={[Function]}
        />
      </WithTheme(CardComponent)>
    </View>
    <TouchableWithoutFeedback
      onPress={[Function]}
      style={
        {
          "justifyContent": "flex-end",
        }
      }
    >
      <View>
        <View
          style={
            {
              "flexDirection": "row",
              "marginBottom": 22,
              "marginTop": 11,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
              }
            }
          >
            <WithTheme(EdgeTextComponent)
              style={
                {
                  "color": "#A4C7DF",
                }
              }
            >
              Fee
            </WithTheme(EdgeTextComponent)>
            <Icon
              allowFontScaling={false}
              name="edit"
              size={17}
              style={
                {
                  "color": "#00f1a2",
                  "marginLeft": 11,
                }
              }
            />
          </View>
          <WithTheme(EdgeTextComponent)
            style={
              {
                "color": "#FFFFFF",
              }
            }
          >
            1 BTC 
            (
            <FiatText
              nativeCryptoAmount="1"
              wallet={{}}
            />
            )
          </WithTheme(EdgeTextComponent)>
        </View>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <WithTheme(EdgeTextComponent)
            style={
              {
                "color": "#A4C7DF",
                "flex": 1,
                "fontSize": 17,
              }
            }
          >
            Rate
          </WithTheme(EdgeTextComponent)>
          <Memo(Connect(WithTheme(ExchangeRateComponent)))
            primaryInfo={
              {
                "displayCurrencyCode": "BTC",
                "displayDenomination": {
                  "multiplier": "100000000000",
                  "name": "BTC",
                },
                "exchangeCurrencyCode": "BTC",
                "exchangeDenomination": {
                  "multiplier": "100000000000",
                  "name": "BTC",
                },
                "tokenId": undefined,
              }
            }
            secondaryDisplayAmount="1000"
            secondaryInfo={
              {
                "displayCurrencyCode": "BTC",
                "displayDenomination": {
                  "multiplier": "1",
                  "name": "Bitcoin",
                },
                "exchangeCurrencyCode": "BTC",
                "exchangeDenomination": {
                  "multiplier": "1",
                  "name": "Bitcoin",
                },
              }
            }
            style={
              {
                "fontSize": 17,
              }
            }
          />
        </View>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <WithTheme(EdgeTextComponent)
            style={
              {
                "color": "#A4C7DF",
                "flex": 1,
                "fontSize": 17,
              }
            }
          >
            Balance
          </WithTheme(EdgeTextComponent)>
          <WithTheme(EdgeTextComponent)
            style={
              {
                "fontSize": 17,
              }
            }
          >
            0.0000001 BTC 
            (
            <FiatText
              nativeCryptoAmount="10000"
              wallet={{}}
            />
            )
          </WithTheme(EdgeTextComponent)>
        </View>
        <WithTheme(EdgeTextComponent)
          numberOfLines={1}
          style={
            [
              {
                "color": "#E85466",
                "fontSize": 17,
              },
              {
                "opacity": 0,
              },
            ]
          }
        >
           
        </WithTheme(EdgeTextComponent)>
      </View>
    </TouchableWithoutFeedback>
  </View>
</ThemedModal>
`;
